pytket/pytket/rmbqc/mpattern.py

```diff
@@ -0,0 +1,533 @@
+        of measurements and corrections which ensures that the underlying
+        graphs are implemented deterministically.
+        
+        :param glist:   A list of unconnected graphs.
+        :param type:    List[GraphS]
+        
+        :returns:       A circuit containing measurements and conditional gates.
+        :rtype:         Circuit
+        """
+        signals: Dict[int, Dict[str, Set[Union[int, bool]]]] = {}
+        total_v = 0
+        for g in glist:
+            total_v += len(g.vertices())
+            for v in g.vertices():
+                signals[v] = {"x": set(), "z": set()}
+        new_c = Circuit(total_v, total_v)
+        for g in glist:
+            gf = gflow(g)
+            if gflow == None:
+                return None
+            else:
+                l_list = MPattern.layer_list(gf[0])
+                layer_num = len(l_list)
+                reset_list: List[int] = []
+                for corr_layer in range(layer_num - 1):
+                    if corr_layer > 0:
+                        isClifford = True
+                        for v in l_list[-1 - corr_layer]:
+                            if not (g.phase(v) in {0, 1 / 2, 1, 3 / 2}):
+                                isClifford = False
+                                break
+                        if not isClifford:
+                            # Uncomment the following command for barriers between layers.
+                            # new_c.add_barrier(list(g.vertices()),list(g.vertices()))
+                            for v in reset_list:
+                                new_c.add_gate(OpType.Reset, [v])
+                            reset_list = []
+                    for v in l_list[-1 - corr_layer]:
+                        my_result = {v}
+                        if g.phase(v) in {0, 1 / 2, 1, 3 / 2}:
+                            my_result ^= signals[v]["z"]
+                        if g.phase(v) in {1 / 2, 1, 3 / 2}:
+                            my_result ^= signals[v]["x"]
+                        if g.phase(v) in {1 / 2, 3 / 2}:
+                            my_result ^= {True}
+                        for u in gf[1][v] - {v}:
+                            signals[u]["x"] ^= my_result
+                        for u in g.vertices() - {v}:
+                            Nu = set(g.neighbors(u))
+                            if (len(Nu & gf[1][v]) % 2) == 1:
+                                signals[u]["z"] ^= my_result
+                        if g.phase(v) in {0, 1}:
+                            new_c.H(v)
+                        elif g.phase(v) in {1 / 2, 3 / 2}:
+                            new_c.Rx(-g.phase(v), v)
+                        else:
+                            new_c.H(v)
+                            # theta = zi-(((-1)**xi)*g.phase(v))
+                            zi = False
+                            for val in signals[v]["z"]:
+                                if type(val) == bool:
```

Comment:
it can be a boolean or an integer (which is reference to some qubit measurement)...

Basically the reason it's written like this it's because it's really convenient to represent these signals as Set[Union[int, bool]]

For example xi = {3,4,9,True} indicates that you get an X correction on some qubit which is based on the measurements on qubits 3, 4, 9 and flipped. The reason I'm using a set is because these dependencies are constantly updated..

For example let's say a correction is originally dependent on {3,4,9,True}, but then we gain another dependency on a measurement which is itself dependent on {3,4,True}. The final correction can be obtained by doing XOR({3,4,9,True},{3,4,True}) = {9}...

So I don't know if this makes sense but that's why I stuck with Set[Union[bool,int]] for these signals and later on I just have to separate the bools from the ints when generating the ClassicalExpBoxes